test/testenv/secretutil.go

```diff
@@ -0,0 +1,133 @@
+	HecToken     string `json:"hec_token"`
+	IdxcSecret   string `json:"idxc_secret"`
+	Pass4SymmKey string `json:"pass4SymmKey"`
+	Password     string `json:"password"`
+	ShcSecret    string `json:"shc_secret"`
+}
+
+// DecodeBase64 decodes base64 and returns string
+func DecodeBase64(str string) string {
+	out, err := b64.StdEncoding.DecodeString(str)
+	if err != nil {
+		logf.Log.Error(err, "Failed to decode", "string", str)
+		return ""
+	}
+	return string(out)
+}
+
+// EncodeBase64 Encodes base64 and returns string
+func EncodeBase64(str string) string {
+	out := b64.StdEncoding.EncodeToString([]byte(str))
+	return out
+}
+
+// GetSecretObject Gets the secret object
+func GetSecretObject(deployment *Deployment, ns string) *SecretResponse {
+	secretObjectName := fmt.Sprintf(SecretObject, ns)
+	output, err := exec.Command("kubectl", "get", "secret", secretObjectName, "-n", ns, "-o", "jsonpath='{.data}'").Output()
+	if err != nil {
+		cmd := fmt.Sprintf("kubectl get secret %s -n %s -o jsonpath='{.data}'", secretObjectName, ns)
+		logf.Log.Error(err, "Failed to execute command", "command", cmd)
+		return nil
+	}
+	// Parse response into response struct
+	restResponse := SecretResponse{}
+	err = json.Unmarshal([]byte(strings.Trim(string(output), "'")), &restResponse)
+	if err != nil {
+		logf.Log.Error(err, "Failed to parse response")
+		return nil
+	}
+	return &restResponse
+}
+
+// GetSecretKey Gets the value to specific key from secret object
+func GetSecretKey(deployment *Deployment, ns string, key string) string {
+	restResponse := GetSecretObject(deployment, ns)
+	//return key based on request
+	switch key {
+	case "hec_token":
+		key := DecodeBase64(restResponse.HecToken)
+		return key
+	case "idxc_secret":
+		key := DecodeBase64(restResponse.IdxcSecret)
+		return key
+	case "pass4SymmKey":
+		key := DecodeBase64(restResponse.Pass4SymmKey)
+		return key
+	case "password":
+		key := DecodeBase64(restResponse.Password)
+		return key
+	case "shc_secret":
+		key := DecodeBase64(restResponse.ShcSecret)
+		return key
+	default:
+		return "Invalid Key"
+	}
+}
+
+//ModifySecretObject Modifies the entire secret object
+func ModifySecretObject(deployment *Deployment, data map[string][]byte, ns string) bool {
+	secretName := fmt.Sprintf(SecretObject, ns)
+	secret := newSecretSpec(ns, secretName, data)
+	//Update object using spec
+	err := deployment.updateCR(secret)
+	if err != nil {
+		logf.Log.Error(err, "Unable to update secret object")
+		return false
+	}
+	return true
+
+}
+
+//ModifySecretKey Modifies the specific key in secret object
+func ModifySecretKey(deployment *Deployment, ns string, key string, value string) bool {
+	//Get current config for update
+	restResponse := GetSecretObject(deployment, ns)
+	out, err := json.Marshal(restResponse)
+	if err != nil {
+		logf.Log.Error(err, "Failed to parse response")
+		return false
+	}
+
+	//Convert object to map for update
+	var data map[string][]byte
+	err = json.Unmarshal([]byte(out), &data)
```

Comment:
Curious, how does the unmarshal convert the variable **out** to a **map**? And what data type is **out**?

Reply:
Out is a Json object. Unmarshal will convert it to a map. Tested it out when running the local changes

Reply:
Does the input to unmarshal need to be of a certain type for it to convert it to a map?

Reply:
So the object is initially built into a json out of the exec.command() function using the struct defined at the top. Here we take care of any extra fields that might have come accidentally.  Post that its simple to convert from json to map and then to update